source/fuzz/transformation_duplicate_region_with_selection.cpp

```diff
@@ -0,0 +1,566 @@
+      fuzzerutil::RepeatedUInt32PairToMap(
+          message_.original_label_to_duplicate_label());
+
+  std::map<uint32_t, uint32_t> original_id_to_duplicate_id =
+      fuzzerutil::RepeatedUInt32PairToMap(
+          message_.original_id_to_duplicate_id());
+
+  std::map<uint32_t, uint32_t> original_id_to_phi_id =
+      fuzzerutil::RepeatedUInt32PairToMap(message_.original_id_to_phi_id());
+
+  for (auto block : region_set) {
+    auto label =
+        ir_context->get_def_use_mgr()->GetDef(block->id())->result_id();
+    // The label of every block in the region must be present in the map
+    // |original_label_to_duplicate_label|.
+    if (original_label_to_duplicate_label.count(label) == 0) {
+      return false;
+    }
+    auto duplicate_label = original_label_to_duplicate_label[label];
+    // Each id assigned to labels in the region must be distinct and fresh.
+    if (!duplicate_label ||
+        !CheckIdIsFreshAndNotUsedByThisTransformation(
+            duplicate_label, ir_context, &ids_used_by_this_transformation)) {
+      return false;
+    }
+    for (auto instr : *block) {
+      if (!instr.HasResultId()) {
+        continue;
+      }
+      // Every instruction with a result id in the region must be present in the
+      // map |original_id_to_duplicate_id|.
+      if (original_id_to_duplicate_id.count(instr.result_id()) == 0) {
+        return false;
+      }
+      auto duplicate_id = original_id_to_duplicate_id[instr.result_id()];
+      // Id assigned to this result id in the region must be distinct and fresh.
+      if (!duplicate_id ||
+          !CheckIdIsFreshAndNotUsedByThisTransformation(
+              duplicate_id, ir_context, &ids_used_by_this_transformation)) {
+        return false;
+      }
+      if (&instr == &*exit_block->tail() ||
+          fuzzerutil::IdIsAvailableBeforeInstruction(
+              ir_context, &*exit_block->tail(), instr.result_id())) {
+        // Every instruction with a result id available at the end of the region
+        // must be present in the map |original_id_to_phi_id|.
+        if (original_id_to_phi_id.count(instr.result_id()) == 0) {
+          return false;
+        }
+        // Using pointers with OpPhi requires capability VariablePointers.
+        if (ir_context->get_type_mgr()->GetType(instr.type_id())->AsPointer() &&
+            !ir_context->get_feature_mgr()->HasCapability(
+                SpvCapabilityVariablePointers)) {
+          return false;
+        }
+        auto phi_id = original_id_to_phi_id[instr.result_id()];
+        // Id assigned to this result id in the region must be distinct and
+        // fresh.
+        if (!phi_id ||
+            !CheckIdIsFreshAndNotUsedByThisTransformation(
+                phi_id, ir_context, &ids_used_by_this_transformation)) {
+          return false;
+        }
+      }
+    }
+  }
+  return true;
+}
+
+std::set<opt::BasicBlock*>
+TransformationDuplicateRegionWithSelection::GetRegionBlocks(
```

Comment:
Minor, but I suggest you move this method to after `Apply`, so that we have `IsApplicable`, then `Apply`, and then helper methods.

Reply:
Also: I think this could be factored out as it duplicates code in TransformationOutlineFunction, so when you open the issue regarding refactoring to avoid that duplication can you put a TODO(...) here as well?